### PR TITLE
Make HEBO sampler supports Define-by-Run manner, maximization, parallelization, and constant_liar

### PR DESCRIPTION
**Pull request by eukaryo**
## Contributor Agreements

Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.

- [x] I agree to the contributor agreements.

> [!TIP]
> Please follow the [Quick TODO list](https://github.com/optuna/optunahub-registry/tree/main?tab=readme-ov-file#quick-todo-list-towards-contribution) to smoothly merge your PR.

## Motivation

I would like to see a HEBO implementation that can be a drop-in replacement of Samplers in Optuna code that is written in Define-by-Run (its search space is not conditional).

I also want it to support parallelization.

## Description of the changes

Modified existing HEBO package.

## TODO List towards PR Merge

Please remove this section if this PR is not an addition of a new package.
Otherwise, please check the following TODO list:

- [x] Copy `./template/` to create your package
- [x] Replace `<COPYRIGHT HOLDER>` in `LICENSE` of your package with your name
- [x] Fill out `README.md` in your package
- [x] Add import statements of your function or class names to be used in `__init__.py`
- [x] (Optional) Add `from __future__ import annotations` at the head of any Python files that include typing to support older Python versions
- [x] Apply the formatter based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)
- [x] Check whether your module works as intended based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)

### CODE REVIEW

**Comment by eukaryo** [comment]
https://github.com/optuna/optunahub-registry/pull/195
In the previous PR, I tried making another package, and some reviews were done.

**Comment by y0z** [comment]
I've updated my comment (https://github.com/optuna/optunahub-registry/pull/196#issuecomment-2516506798).

**Review by y0z — APPROVED** [review]
LGTM

@nabenabe0928 If you have no further concerns, please merge this PR.

**Comment by eukaryo** [comment]
Note: I added "stateless" mode for parallelization; in the mode, it creates entire HEBO model in every trial and thus requires more computational cost. I preserved previous "define-and-run" mode for compartibility. It is not stateless but faster.

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:108]: Followed TPESampler
```suggestion
        use_cache = not self._constant_liar
        trials = study._get_trials(deepcopy=False, states=states, use_cache=use_cache)
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:110]:
```suggestion
            # note: The backend HEBO implementation uses Sobol sampling here.
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:119]:
```suggestion
            worst_value = max(t.value for t in trials if t.state == TrialState.COMPLETE)
        else:
            worst_value = min(t.value for t in trials if t.state == TrialState.COMPLETE)
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:88]:
```suggestion
        self._constant_liar = constant_liar
        self._rng = np.random.default_rng(seed)
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:122]:
```suggestion
        seed = self._rng.randint((1 << 31) - 1)
        hebo = HEBO(self._convert_to_hebo_design_space(search_space), scramble_seed=seed)
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:146]:
```suggestion
        return {name: row[0] for name, row in hebo.suggest().items()}
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:141]:
```suggestion
        valid_trials = [t.params for t in trials if all(name in trial.params for name in search_space)]
        params = pd.DataFrame([t.params for t in valid_trials])
        values = np.array([sign * t.value if t.state == TrialState.COMPLETE else worst_value for t in valid_trials])
        hebo.observe(params, values)
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:76]:
```suggestion
        *,
        seed: int | None = None,
```

File: package/samplers/hebo/README.md
Comment (nabenabe0928) [inline package/samplers/hebo/README.md:25]:
```suggestion
- `HEBOSampler(search_space: dict[str, BaseDistribution] | None = None, *, seed: int | None = None, constant_liar: bool = False, independent_sampler: BaseSampler | None = None)`
```

File: package/samplers/hebo/README.md
Comment (nabenabe0928) [inline package/samplers/hebo/README.md:26]:
```suggestion
  - `search_space`: By specifying search_space, the sampling speed at each iteration becomes slightly quicker, but this argument is not necessary to run this sampler.
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:34]:
```suggestion
            By specifying search_space, the sampling speed at each iteration becomes slightly quicker, but this argument is not necessary to run this sampler. Default is :obj:`None`.
```

File: package/samplers/hebo/README.md
Comment (nabenabe0928) [inline package/samplers/hebo/README.md:65]:
```suggestion
sampler = module.HEBOSampler()
```

**Review by nabenabe0928 — COMMENTED** [review]
Thank you for the PR, I left some comments:)

File: package/samplers/hebo/sampler.py
Comment (eukaryo) [reply package/samplers/hebo/sampler.py]: Agreed. I revised the code. (The warn is not needed iff it is the first trial of a define-by-run mode; it is considered in the revision.)

File: package/samplers/hebo/sampler.py
Comment (eukaryo) [reply package/samplers/hebo/sampler.py]: Basically agreed. I revised the code to `return {name: row.iloc[0] for name, row in self._hebo.suggest().items() if name in search_space.keys()}`.

File: package/samplers/hebo/README.md
Comment (eukaryo) [reply package/samplers/hebo/README.md:65]: Basically agreed. `search_space` is not required but acceptable. So I added a commented line, `# sampler = module.HEBOSampler()  # Note: search_space is not required, and thus it works too.`.

**Comment by eukaryo** [comment]
@nabenabe0928 Thank you for comments! I revised the code. PTAL.

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:248]: Probably, warning happens when fallback is inevitable, but when it is NOT inevitable?

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:132]:
```suggestion
        params = pd.DataFrame([
            t.params for t in trials if all(name in trial.params for name in search_space)
        ])
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:125]: Probably, `self._is_independent_sample_necessary` is easier to understand.

```suggestion
        is_complete = np.array([t.state == TrialState.COMPLETE for t in trials])
        if not np.any(is_complete):
            # note: The backend HEBO implementation uses Sobol sampling here.
            # This sampler does not call `hebo.suggest()` here because
            # Optuna needs to know search space by running the first trial in Define-by-Run.
            self._is_independent_sample_necessary = True
            return {}
        else:
            self._is_independent_sample_necessary = False

        # Assume that the back-end HEBO implementation aims to minimize.
        values = np.array([t.value if t.state == TrialState.COMPLETE else np.nan for t in trials])
        worst_value = np.nanmax(values) if study.direction == StudyDirection.MINIMIZE else np.nanmin(values)
        sign = 1 if study.direction == StudyDirection.MINIMIZE else -1
```

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:138]:
```suggestion
        values[np.isnan(values)] = worst_value
        values *= sign
```

**Review by nabenabe0928 — COMMENTED** [review]
I left some comments, but you do not have to address all!
In the worst case, we can once merge this PR and can make another PR!

**Comment by eukaryo** [comment]
Thank you for additional comments!
I checked the current code works in the following senarios: (1)`search_space` is given (2) `search_space=None` (3) `search_space=None, constant_liar=True`, and `n_jobs=3`.
@nabenabe0928 would you please merge this PR if you have no further critical concerns?

File: package/samplers/hebo/sampler.py
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:134]: I will check if the shape is correct here (seems `(1, N)` is correct?)
Anyways, I will benchmark your code and merge this PR!

**Review by nabenabe0928 — COMMENTED** [review]
LGTM, but I will check if the performance is fine just in case!

File: package/samplers/hebo/sampler.py
Comment (eukaryo) [reply package/samplers/hebo/sampler.py:134]: Thanks! fyi: I confirmed that simply changing line134 to `hebo.observe(params, values[None, :])` fails.

**Comment by nabenabe0928** [comment]
I confirmed it works!

**Pull request merged**
